docs/imds_retry_based_on_errors.md

```diff
@@ -0,0 +1,67 @@
+|----------------------|-----------------------------------------------|---------------------------------------------|-----------------------------------------|
+| **400**             | Bad Request (Invalid Parameters)               | **Do not retry**, fix request               | **No retry**                             |
+| **401**             | Unauthorized                                  | **Do not retry**, check authentication setup | **No retry**                             |
+| **403**             | Forbidden                                     | **Do not retry**, verify permissions       | **No retry**                             |
+| **404**             | IMDS endpoint is updating / Identity Not Found | Retry with Exponential Backoff (max 3 retries) | **1s → 2s → 4s (max 4s)**            |
+| **408**             | Request Timeout                                | Retry with Exponential Backoff (max 3 retries) | **1s → 2s → 4s (max 4s)**            |
+| **410**             | IMDS is undergoing updates                    | Retry with Exponential Backoff (max 3 retries) | **1s → 2s → 4s (max 4s)**            |
+| **429**             | IMDS Throttle limit reached                   | Retry with Exponential Backoff (max 3 retries) | **1s → 2s → 4s (max 4s)**            |
+| **504**             | Gateway Timeout                               | Retry with Exponential Backoff (max 3 retries) | **1s → 2s → 4s (max 4s)**            |
+| **5xx**             | Transient service error                        | Retry with Exponential Backoff (max 3 retries) | **1s → 2s → 4s (max 4s)**            |
+
+---
+
+## 2️⃣ Identity Propagation & Special Handling for "Identity Not Found" Errors
+- **Scenario:** When an identity is newly assigned to a VM, it may take time for the IMDS service to recognize the identity.
+- **Exception Handling:**  
+  - If the **IMDS response contains "Identity Not Found"**, retry the request using **exponential backoff**.
+  - **Error Code:** **404 (Identity Not Found)**
+  - Recommended retry sequence: **1s → 2s → 4s** (max 3 retries)
+  - If still failing, log an error and return the failure.
+
+---
+
+## 3️⃣ Updated Retry Strategy (Exponential Backoff)
+The following retry strategy applies to **5xx errors, timeouts, and transient 4xx errors (e.g., Identity Not Found):**
+
+| **Retry Attempt** | **Delay Before Retry** |
+|------------------|----------------------|
+| **1st**         | **1 second**         |
+| **2nd**         | **2 seconds**         |
+| **3rd**         | **4 seconds** (max 4s) |
+
+🔹 **For 5xx Errors, 404 Identity Not Found, and Timeouts:** Retry **max 3 times** before failing.  
+🔹 **For 410 (IMDS Updates):** Backoff **increases on each retry** (1s → 2s → 4s - max 4s).
+🔹 **For 429 (Throttling):** Backoff **increases on each retry** (1s → 2s → 4s - max 4s).
+
+---
+
+```mermaid
```

Comment:
How can I see this graph for the purposes of a review?

Reply:
rich diff should show the diagram. If the Mermaid diagram does not render automatically, copy it into [mermaid.live](https://mermaid.live/).